quickwit/quickwit-search/src/leaf.rs

```diff
@@ -829,6 +829,104 @@ impl CanSplitDoBetter {
     }
 }
 
+/// `multi_leaf_search` searches multiple indices and multiple splits.
+#[instrument(skip_all, fields(index = ?leaf_search_request.search_request.as_ref().unwrap().index_id_patterns))]
+pub async fn multi_leaf_search(
+    searcher_context: Arc<SearcherContext>,
+    leaf_search_request: LeafSearchRequest,
+    storage_resolver: &StorageResolver,
+) -> Result<LeafSearchResponse, SearchError> {
+    let search_request: Arc<SearchRequest> = leaf_search_request
+        .search_request
+        .ok_or_else(|| SearchError::Internal("no search request".to_string()))?
+        .into();
+
+    let doc_mappers: Vec<Arc<dyn DocMapper>> = leaf_search_request
+        .doc_mappers
+        .iter()
+        .map(|doc_mapper| deserialize_doc_mapper(doc_mapper))
+        .collect::<crate::Result<_>>()?;
+    // Creates a collector which merges responses into one
+    let aggregation_limits = searcher_context.create_new_aggregation_limits();
+    // TODO: to avoid lockstep, we should pull up the future creation over the list of split ids
+    // and have the semaphore on this level.
+    // This will lower resource consumption due to less in-flight futures and avoid contention.
+    // It also allows passing early exit conditions between indices.
+    //
+    // It is a little bit tricky how to handle which is now the incremental_merge_collector, one
+    // per index, e.g. when to merge results and how to avoid lock contention.
+    let mut leaf_request_tasks = Vec::new();
+
+    for leaf_search_request_ref in leaf_search_request.leaf_requests.into_iter() {
+        let index_uri = quickwit_common::uri::Uri::from_str(
+            &leaf_search_request.index_uris[leaf_search_request_ref.index_uri_ord as usize],
+        )?;
+        let doc_mapper = doc_mappers[leaf_search_request_ref.doc_mapper_ord as usize].clone();
```

Comment:
```suggestion
        let doc_mapper = doc_mappers[leaf_search_request_ref.doc_mapper_ord as usize].clone();
```
```suggestion
        let doc_mapper = doc_mappers.get(leaf_search_request_ref.doc_mapper_ord as usize).map_err(...)?.clone();
```

The point is that here the contract of having doc_mapper_ord referring to a properly enforced doc_mappers array is enforced by the grpc client that ran the request. This is code from a different file, running on a different server, possibly from a different version of quickwit, possibly not even coming from quickwit.

It is not something we want to panic on. We should return a proper error here.